crates/primitives/utils/src/service.rs

```diff
@@ -1,21 +1,297 @@
+                    _ = self.token_global.cancelled() => {},
+                },
+            }
+        }
+    }
+
+    #[inline(always)]
+    pub fn is_cancelled(&self) -> bool {
+        self.token_global.is_cancelled()
+            || self.token_local.as_ref().map(|t| t.is_cancelled()).unwrap_or(false)
+            || !self.capabilities.is_active(self.id as u8)
+            || self.state() == MadaraState::Shutdown
+    }
+
+    pub fn id(&self) -> MadaraCapability {
+        self.id
+    }
+
+    /// Copies the context, maintaining its scope.
+    pub fn branch(&self) -> Self {
+        Self {
+            token_global: self.token_global.clone(),
+            token_local: self.token_local.clone(),
+            capabilities: Arc::clone(&self.capabilities),
+            capabilities_notify: Arc::clone(&self.capabilities_notify),
+            state: Arc::clone(&self.state),
+            id: self.id,
+        }
+    }
+
+    /// Copies the context, maintaining its scope but with a new id.
+    pub fn branch_id(&self, id: MadaraCapability) -> Self {
+        Self {
+            token_global: self.token_global.clone(),
+            token_local: self.token_local.clone(),
+            capabilities: Arc::clone(&self.capabilities),
+            capabilities_notify: Arc::clone(&self.capabilities_notify),
+            state: Arc::clone(&self.state),
+            id,
+        }
+    }
+
+    /// Copies the context into a new local scope.
+    ///
+    /// Any service which uses this new context will be able to cancel the
+    /// services in the same local scope as itself, and any further child
+    /// services, without affecting the rest of the global scope.
+    pub fn child(&self, id: MadaraCapability) -> Self {
+        let token_local = self.token_local.as_ref().unwrap_or(&self.token_global).child_token();
+
+        Self {
+            token_global: self.token_global.clone(),
+            token_local: Some(token_local),
+            capabilities: Arc::clone(&self.capabilities),
+            capabilities_notify: Arc::clone(&self.capabilities_notify),
+            state: Arc::clone(&self.state),
+            id,
+        }
+    }
+
+    /// Atomically checks if a set of services are running.
+    ///
+    /// You can combine multiple [MadaraCapability] into a single bitmask to
+    /// check the state of multiple services at once.
+    #[inline(always)]
+    pub fn capabilities_check(&self, cap: u8) -> bool {
+        self.capabilities.is_active(cap)
+    }
+
+    /// Atomically marks a service as active
+    ///
+    /// This will immediately be visible to all services in the same global
+    /// scope. This is true across threads.
+    #[inline(always)]
+    pub fn capabilities_add(&self, cap: MadaraCapability) -> bool {
+        let res = self.capabilities.activate(cap);
+        self.capabilities_notify.notify_waiters();
+
+        res
+    }
+
+    #[inline(always)]
+    pub fn capabilities_remove(&self, cap: MadaraCapability) -> bool {
```

Comment:
i dont like that name, it doesnt convey that this disables a service

Reply:
This has been fixed